File: src/sample.c
/*
 * sample.c - Sample subcommand for depth-aware BAM sampling
 *
 * This module samples reads from a source BAM file based on depth
 * distribution from one or more template BED files. Reads are selected
 * using a deterministic hash-based approach that allows for reproducible
 * sampling.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <getopt.h>

#include <htslib/sam.h>
#include <htslib/hts.h>

#include "sample.h"
#include "bed.h"
#include "depth.h"
#include "metrics.h"
#include "xxhash.h"
#include "samsampleX.h"

/*
 * Print sample subcommand usage.
 */
void sample_usage(void) {
    fprintf(stderr, "Usage: %s sample [options]\n\n", SAMSAMPLEX_NAME);
    fprintf(stderr, "Sample reads from a BAM file to match depth distribution from template BED(s).\n\n");
    fprintf(stderr, "Required options:\n");
    fprintf(stderr, "  --source-bam FILE     Input BAM file to sample reads from\n");
    fprintf(stderr, "  --template-bed FILE   Template BED file(s) with depth values (can be repeated)\n");
    fprintf(stderr, "  --region REGION       Target region (samtools-style)\n\n");
    fprintf(stderr, "Optional:\n");
    fprintf(stderr, "  --out-bam FILE        Output BAM file [default: %s]\n", DEFAULT_OUT_BAM);
    fprintf(stderr, "  --mode MODE           How to combine multiple templates: min, max, mean, random\n");
    fprintf(stderr, "                        [default: min]\n");
    fprintf(stderr, "  --seed INT            Random seed for deterministic sampling [default: %d]\n", DEFAULT_SEED);
    fprintf(stderr, "  --no-sort             Do not sort and index output BAM file\n");
    fprintf(stderr, "  -h, --help            Show this help message\n");
}

/*
 * Compute cumulative sum array from ratio array.
 * Returns array of size (length + 1) where cumsum[0] = 0.
 */
static double *compute_cumsum(const double *ratios, size_t length) {
    double *cumsum;
    SAFE_MALLOC(cumsum, (length + 1) * sizeof(double));
    
    cumsum[0] = 0.0;
    for (size_t i = 0; i < length; i++) {
        cumsum[i + 1] = cumsum[i] + ratios[i];
    }
    
    return cumsum;
}

/*
 * Get mean ratio for a read spanning [read_start, read_end) using cumulative sum.
 */
static double get_mean_ratio(const double *cumsum, int64_t region_start,
                             int64_t region_end, int64_t read_start, int64_t read_end) {
    /* Clip read coordinates to region */
    int64_t clip_start = MAX(read_start, region_start);
    int64_t clip_end = MIN(read_end, region_end);
    
    if (clip_start >= clip_end) {
        return 0.0;
    }
    
    /* Convert to array indices */
    size_t i1 = (size_t)(clip_start - region_start);
    size_t i2 = (size_t)(clip_end - region_start);
    
    double sum = cumsum[i2] - cumsum[i1];
    return sum / (double)(clip_end - clip_start);
}

/*
 * Run the sample subcommand.
 */
int sample_run(sample_args_t *args) {
    int ret = 0;
    
    /* Validate required arguments */
    if (!args->source_bam) {
        fprintf(stderr, "Error: --source-bam is required\n");
        sample_usage();
        return 1;
    }
    if (args->n_template_beds == 0) {
        fprintf(stderr, "Error: At least one --template-bed is required\n");
        sample_usage();
        return 1;
    }
    if (!args->region) {
        fprintf(stderr, "Error: --region is required\n");
        sample_usage();
        return 1;
    }
    
    /* Set defaults */
    if (!args->out_bam) {
        args->out_bam = DEFAULT_OUT_BAM;
    }
    
    fprintf(stderr, "[sample] Source BAM: %s\n", args->source_bam);
    fprintf(stderr, "[sample] Template BEDs: %d file(s)\n", args->n_template_beds);
    for (int i = 0; i < args->n_template_beds; i++) {
        fprintf(stderr, "[sample]   %d: %s\n", i + 1, args->template_beds[i]);
    }
    fprintf(stderr, "[sample] Region: %s\n", args->region);
    fprintf(stderr, "[sample] Mode: %s\n", combine_mode_to_string(args->mode));
    fprintf(stderr, "[sample] Seed: %u\n", args->seed);
    fprintf(stderr, "[sample] Output BAM: %s\n", args->out_bam);
    
    /* Parse region */
    region_t *region = region_parse(args->region);
    if (!region || !region->contig) {
        fprintf(stderr, "Error: Invalid region format: %s\n", args->region);
        return 1;
    }
    
    /* Open source BAM to get header and determine region bounds */
    htsFile *source_fp = hts_open(args->source_bam, "r");
    if (!source_fp) {
        fprintf(stderr, "Error: Cannot open source BAM: %s\n", args->source_bam);
        region_free(region);
        return 1;
    }
    
    sam_hdr_t *header = sam_hdr_read(source_fp);
    if (!header) {
        fprintf(stderr, "Error: Cannot read BAM header\n");
        hts_close(source_fp);
        region_free(region);
        return 1;
    }
    
    /* Load index */
    hts_idx_t *idx = sam_index_load(source_fp, args->source_bam);
    if (!idx) {
        fprintf(stderr, "Error: Cannot load BAM index. Make sure .bai file exists.\n");
        sam_hdr_destroy(header);
        hts_close(source_fp);
        region_free(region);
        return 1;
    }
    
    /* Get contig ID and length */
    int tid = sam_hdr_name2tid(header, region->contig);
    if (tid < 0) {
        fprintf(stderr, "Error: Contig '%s' not found in BAM\n", region->contig);
        hts_idx_destroy(idx);
        sam_hdr_destroy(header);
        hts_close(source_fp);
        region_free(region);
        return 1;
    }
    
    /* Set region bounds */
    if (region->start < 0) region->start = 0;
    if (region->end < 0) region->end = sam_hdr_tid2len(header, tid);
    
    fprintf(stderr, "[sample] Parsed region: %s:%ld-%ld\n",
            region->contig, (long)region->start, (long)region->end);
    
    /* Load template depth arrays from BED files */
    fprintf(stderr, "[sample] Loading template BED file(s)...\n");
    depth_array_t *template_arrays[MAX_TEMPLATE_BEDS];
    for (int i = 0; i < args->n_template_beds; i++) {
        template_arrays[i] = bed_read_depths(args->template_beds[i], region->contig,
                                             region->start, region->end);
        if (!template_arrays[i]) {
            fprintf(stderr, "Error: Failed to load BED file: %s\n", args->template_beds[i]);
            for (int j = 0; j < i; j++) {
                depth_array_free(template_arrays[j]);
            }
            hts_idx_destroy(idx);
            sam_hdr_destroy(header);
            hts_close(source_fp);
            region_free(region);
            return 1;
        }
    }
    
    /* Combine template depths */
    depth_array_t *template_depth;
    if (args->n_template_beds == 1) {
        template_depth = template_arrays[0];
    } else {
        fprintf(stderr, "[sample] Combining %d templates using '%s' mode...\n",
                args->n_template_beds, combine_mode_to_string(args->mode));
        template_depth = bed_combine_depths(template_arrays, args->n_template_beds,
                                            args->mode, args->seed);
        /* Free individual arrays */
        for (int i = 0; i < args->n_template_beds; i++) {
            depth_array_free(template_arrays[i]);
        }
        if (!template_depth) {
            fprintf(stderr, "Error: Failed to combine template depths\n");
            hts_idx_destroy(idx);
            sam_hdr_destroy(header);
            hts_close(source_fp);
            region_free(region);
            return 1;
        }
    }
    
    /* Compute source depth */
    fprintf(stderr, "[sample] Computing source depth array...\n");
    depth_array_t *source_depth = depth_from_bam(args->source_bam, region->contig,
                                                  region->start, region->end);
    if (!source_depth) {
        fprintf(stderr, "Error: Failed to compute source depth\n");
        depth_array_free(template_depth);
        hts_idx_destroy(idx);
        sam_hdr_destroy(header);
        hts_close(source_fp);
        region_free(region);
        return 1;
    }
    
    /* Compute ratio array: ratio[i] = min(1.0, template[i] / source[i]) */
    fprintf(stderr, "[sample] Computing sampling ratios...\n");
    double *ratios;
    SAFE_MALLOC(ratios, source_depth->length * sizeof(double));
    
    for (size_t i = 0; i < source_depth->length; i++) {
        if (source_depth->depths[i] == 0) {
            ratios[i] = 0.0;  /* No reads to sample */
        } else {
            double r = (double)template_depth->depths[i] / (double)source_depth->depths[i];
            ratios[i] = (r > 1.0) ? 1.0 : r;  /* Cap at 1.0 */
        }
    }
    
    /* Compute cumulative sum for efficient range queries */
    double *cumsum = compute_cumsum(ratios, source_depth->length);
    free(ratios);
    
    /* Open output BAM for writing */
    htsFile *out_fp = hts_open(args->out_bam, "wb");
    if (!out_fp) {
        fprintf(stderr, "Error: Cannot open output BAM: %s\n", args->out_bam);
        free(cumsum);
        depth_array_free(source_depth);
        depth_array_free(template_depth);
        hts_idx_destroy(idx);
        sam_hdr_destroy(header);
        hts_close(source_fp);
        region_free(region);
        return 1;
    }
    
    /* Write header */
    if (sam_hdr_write(out_fp, header) < 0) {
        fprintf(stderr, "Error: Failed to write BAM header\n");
        hts_close(out_fp);
        free(cumsum);
        depth_array_free(source_depth);
        depth_array_free(template_depth);
        hts_idx_destroy(idx);
        sam_hdr_destroy(header);
        hts_close(source_fp);
        region_free(region);
        return 1;
    }
    
    /* Iterate through reads and sample */
    fprintf(stderr, "[sample] Sampling reads...\n");
    
    hts_itr_t *iter = sam_itr_queryi(idx, tid, region->start, region->end);
    if (!iter) {
        fprintf(stderr, "Error: Cannot create iterator for region\n");
        ret = 1;
        goto cleanup;
    }
    
    bam1_t *b = bam_init1();
    uint64_t total_reads = 0;
    uint64_t kept_reads = 0;
    
    while (sam_itr_next(source_fp, iter, b) >= 0) {
        /* Skip unmapped reads */
        if (b->core.flag & BAM_FUNMAP) {
            continue;
        }
        
        total_reads++;
        
        /* Get read name and compute hash fraction */
        const char *qname = bam_get_qname(b);
        double hash_fraction = XXH32_fraction(qname, args->seed);
        
        /* Get mean ratio over read's covered positions */
        int64_t read_start = b->core.pos;
        int64_t read_end = bam_endpos(b);
        double mean_ratio = get_mean_ratio(cumsum, region->start, region->end,
                                           read_start, read_end);
        
        /* Keep read if hash_fraction < mean_ratio */
        if (hash_fraction < mean_ratio) {
            if (sam_write1(out_fp, header, b) < 0) {
                fprintf(stderr, "Error: Failed to write read\n");
                ret = 1;
                break;
            }
            kept_reads++;
        }
        
        /* Progress report every 1M reads */
        if (total_reads % 1000000 == 0) {
            fprintf(stderr, "[sample]   Processed %lu reads, kept %lu (%.1f%%)\n",
                    (unsigned long)total_reads, (unsigned long)kept_reads,
                    100.0 * kept_reads / total_reads);
        }
    }
    
    fprintf(stderr, "[sample] Processed %lu reads, kept %lu (%.1f%%)\n",
            (unsigned long)total_reads, (unsigned long)kept_reads,
            total_reads > 0 ? 100.0 * kept_reads / total_reads : 0.0);
    
    bam_destroy1(b);
    hts_itr_destroy(iter);
    
    /* Close files before sorting */
    hts_close(out_fp);
    out_fp = NULL;
    
    /* Sort and index output BAM */
    if (!args->no_sort && ret == 0) {
        fprintf(stderr, "[sample] Sorting output BAM...\n");
        
        /* Create temp file for sorted output */
        char tmp_path[1024];
        snprintf(tmp_path, sizeof(tmp_path), "%s.tmp.bam", args->out_bam);
        
        /* Use samtools sort via htslib */
        char sort_cmd[4096];
        snprintf(sort_cmd, sizeof(sort_cmd), 
                 "samtools sort -o '%s' '%s' && mv '%s' '%s' && samtools index '%s'",
                 tmp_path, args->out_bam, tmp_path, args->out_bam, args->out_bam);
        
        int sort_ret = system(sort_cmd);
        if (sort_ret != 0) {
            fprintf(stderr, "Warning: Failed to sort/index output BAM (exit code %d)\n", sort_ret);
            fprintf(stderr, "         The output BAM may be unsorted.\n");
        } else {
            fprintf(stderr, "[sample] Sorting and indexing complete.\n");
        }
    }
    
    /* Calculate and print metrics */
    if (ret == 0) {
        fprintf(stderr, "[sample] Computing metrics...\n");
        
        /* Get output depth */
        depth_array_t *output_depth = depth_from_bam(args->out_bam, region->contig,
                                                      region->start, region->end);
        if (output_depth) {
            metrics_result_t metrics;
            if (metrics_calculate(template_depth, output_depth, &metrics) == 0) {
                metrics_print(&metrics);
            }
            depth_array_free(output_depth);
        } else {
            fprintf(stderr, "Warning: Could not compute metrics\n");
        }
    }
    
cleanup:
    free(cumsum);
    depth_array_free(source_depth);
    depth_array_free(template_depth);
    hts_idx_destroy(idx);
    sam_hdr_destroy(header);
    hts_close(source_fp);
    if (out_fp) hts_close(out_fp);
    region_free(region);
    
    if (ret == 0) {
        fprintf(stderr, "[sample] Done. Output written to: %s\n", args->out_bam);
    }
    
    return ret;
}

File: samsamplex/sample.py
"""Hash-based deterministic BAM sampling to match a template depth profile."""

from __future__ import annotations

import os
import sys
from typing import Sequence

import numpy as np
import pysam
import xxhash

from .bed import bed_combine_depths, bed_read_depths
from .depth import DepthArray, depth_from_bam, region_parse, resolve_contig_name
from .modes import DEPTH_MODES

UINT32_MAX = 0xFFFFFFFF

# Backwards-compatible aliases (same canonical set as map/sample --mode and --stat).
VALID_STAT_MODES = DEPTH_MODES
VALID_COMBINE_MODES = DEPTH_MODES


def _xxh32_fraction(qname: str, seed: int) -> float:
    """Hash a read name to a float in [0, 1)."""
    h = xxhash.xxh32(qname.encode(), seed=seed).intdigest()
    return h / UINT32_MAX


# ── Ratio helpers ────────────────────────────────────────────────────────────

# CIGAR op codes (BAM spec): only M(0), =(7), X(8) are "aligned" positions where the
# read actually has a base aligned to the reference. D(2), N(3) consume reference
# positions but the read has no base there, so we skip them when aggregating ratios.
# I(1), S(4), H(5), P(6) don't consume reference at all.
_ALIGNED_OPS = {0, 7, 8}        # M, =, X (count toward the stat)
_REF_OPS = {0, 2, 3, 7, 8}      # M, D, N, =, X (consume reference; advance ref_pos)


def _compute_ratios(template: DepthArray, source: DepthArray) -> np.ndarray:
    """ratio[i] = min(1.0, template[i] / source[i]), 0 where source is 0."""
    with np.errstate(divide="ignore", invalid="ignore"):
        ratios = np.where(
            source.depths == 0,
            0.0,
            np.minimum(1.0, template.depths.astype(np.float64) / source.depths.astype(np.float64)),
        )
    return ratios


def _aligned_ref_runs(read) -> list[tuple[int, int]]:
    """Walk a read's CIGAR and return [(ref_start, ref_end), ...] for M/=/X spans.

    D and N ops still advance the reference position (so subsequent aligned ops
    map to the right place) but they themselves are not emitted. I/S/H/P ops
    don't consume reference and are silently skipped.
    """
    runs: list[tuple[int, int]] = []
    cigar = read.cigartuples
    if not cigar:
        return runs
    pos = read.reference_start
    for op, length in cigar:
        if op in _ALIGNED_OPS:
            runs.append((pos, pos + length))
            pos += length
        elif op in _REF_OPS:
            pos += length
    return runs


def _collect_aligned_ratios(
    ratios: np.ndarray,
    region_start: int,
    region_end: int,
    runs: list[tuple[int, int]],
) -> np.ndarray | None:
    """Concatenate the slices of *ratios* corresponding to the clipped aligned runs.

    Returns None if no run overlaps the region (so callers can short-circuit to 0.0).
    """
    parts: list[np.ndarray] = []
    for rs, re in runs:
        cs = max(rs, region_start)
        ce = min(re, region_end)
        if cs < ce:
            parts.append(ratios[cs - region_start : ce - region_start])
    if not parts:
        return None
    return parts[0] if len(parts) == 1 else np.concatenate(parts)


def _get_mean_ratio(
    cumsum: np.ndarray,
    region_start: int,
    region_end: int,
    runs: list[tuple[int, int]],
) -> float:
    """Mean ratio over aligned reference positions, using cumsum per run for speed."""
    total = 0.0
    count = 0
    for rs, re in runs:
        cs = max(rs, region_start)
        ce = min(re, region_end)
        if cs >= ce:
            continue
        total += float(cumsum[ce - region_start] - cumsum[cs - region_start])
        count += ce - cs
    if count == 0:
        return 0.0
    return total / count


def _get_min_ratio(
    ratios: np.ndarray,
    region_start: int,
    region_end: int,
    runs: list[tuple[int, int]],
) -> float:
    arr = _collect_aligned_ratios(ratios, region_start, region_end, runs)
    if arr is None or arr.size == 0:
        return 0.0
    return float(arr.min())


def _get_max_ratio(
    ratios: np.ndarray,
    region_start: int,
    region_end: int,
    runs: list[tuple[int, int]],
) -> float:
    arr = _collect_aligned_ratios(ratios, region_start, region_end, runs)
    if arr is None or arr.size == 0:
        return 0.0
    return float(arr.max())


def _get_median_ratio(
    ratios: np.ndarray,
    region_start: int,
    region_end: int,
    runs: list[tuple[int, int]],
) -> float:
    arr = _collect_aligned_ratios(ratios, region_start, region_end, runs)
    if arr is None or arr.size == 0:
        return 0.0
    return float(np.median(arr))


def _get_random_ratio(
    ratios: np.ndarray,
    region_start: int,
    region_end: int,
    runs: list[tuple[int, int]],
    seed: int,
    read_start: int,
    read_end: int,
) -> float:
    """Pick one ratio from the read's aligned positions; index is deterministic in (span, seed).

    The hash key is the original (read_start, read_end) span so determinism per
    (span, seed) is preserved across the CIGAR-aware change; only the array
    being indexed differs.
    """
    arr = _collect_aligned_ratios(ratios, region_start, region_end, runs)
    if arr is None or arr.size == 0:
        return 0.0
    h = xxhash.xxh32(f"{read_start}:{read_end}".encode(), seed=seed).intdigest()
    return float(arr[h % arr.size])


# ── Main sampling routine ───────────────────────────────────────────────────


def sample_run(
    source_bam: str,
    template_beds: Sequence[str],
    region_str: str,
    out_bam: str = "out.bam",
    mode: str = "random",
    stat: str = "mean",
    seed: int = 42,
    # no_sort: bool = False,
    uniform_fraction: float | None = None,
    threads: int = 2,
) -> int:
    """Run the sample subcommand. Returns 0 on success."""
    log = lambda msg: print(msg, file=sys.stderr)

    log(f"[sample] Source BAM: {source_bam}")
    if uniform_fraction is not None:
        log(f"[sample] Uniform fraction: {uniform_fraction}")
    else:
        log(f"[sample] Template BEDs: {len(template_beds)} file(s)")
        for i, b in enumerate(template_beds):
            log(f"[sample]   {i + 1}: {b}")
        log(f"[sample] Stat: {stat}")
        log(f"[sample] Mode: {mode}")
    log(f"[sample] Region: {region_str}")
    log(f"[sample] Seed: {seed}")
    log(f"[sample] Output BAM: {out_bam}")

    region = region_parse(region_str)

    # Open source BAM to resolve region bounds
    with pysam.AlignmentFile(source_bam, "rb") as src:
        resolved = resolve_contig_name(src.header, region.contig)
        if resolved is None:
            log(f"Error: Contig '{region.contig}' not found in BAM")
            return 1
        region.contig = resolved

        if region.start < 0:
            region.start = 0
        if region.end < 0:
            region.end = src.get_reference_length(resolved)

    log(f"[sample] Parsed region: {region.contig}:{region.start}-{region.end}")

    if uniform_fraction is None:
        if mode not in DEPTH_MODES:
            log(
                f"Error: Unknown combine mode {mode!r} "
                f"(expected one of: {', '.join(DEPTH_MODES)})"
            )
            return 1
        if stat not in DEPTH_MODES:
            log(
                f"Error: Unknown stat mode {stat!r} "
                f"(expected one of: {', '.join(DEPTH_MODES)})"
            )
            return 1
        if not template_beds:
            log("Error: Template BED(s) required when not using --uniform")
            return 1

        # Load template depth(s)
        log("[sample] Loading template BED file(s)...")
        template_arrays = [
            bed_read_depths(bp, region.contig, region.start, region.end)
            for bp in template_beds
        ]

        if len(template_arrays) == 1:
            template_depth = template_arrays[0]
        else:
            log(f"[sample] Combining {len(template_arrays)} templates using '{mode}' mode...")
            template_depth = bed_combine_depths(template_arrays, mode=mode, seed=seed)

        # Compute source depth
        log("[sample] Computing source depth array...")
        source_depth = depth_from_bam(source_bam, region.contig, region.start, region.end)

        # Compute ratios
        log("[sample] Computing sampling ratios...")
        ratios = _compute_ratios(template_depth, source_depth)

        # Precompute cumulative sum (used by mean stat mode, always needed for default)
        cumsum = np.concatenate(([0.0], np.cumsum(ratios)))

        # Choose ratio-lookup function based on stat mode. Each lambda walks the
        # read's CIGAR (via _aligned_ref_runs) so only M/=/X reference positions
        # contribute; D/N/I/S/H/P are excluded.
        if stat == "mean":
            get_ratio = lambda r: _get_mean_ratio(
                cumsum, region.start, region.end, _aligned_ref_runs(r)
            )
        elif stat == "min":
            get_ratio = lambda r: _get_min_ratio(
                ratios, region.start, region.end, _aligned_ref_runs(r)
            )
        elif stat == "max":
            get_ratio = lambda r: _get_max_ratio(
                ratios, region.start, region.end, _aligned_ref_runs(r)
            )
        elif stat == "median":
            get_ratio = lambda r: _get_median_ratio(
                ratios, region.start, region.end, _aligned_ref_runs(r)
            )
        elif stat == "random":
            get_ratio = lambda r: _get_random_ratio(
                ratios, region.start, region.end, _aligned_ref_runs(r),
                seed, r.reference_start, r.reference_end,
            )
        else:
            log(f"Error: Unknown stat mode '{stat}'")
            return 1

    # Sampling loop
    log("[sample] Sampling reads...")
    total_reads = 0
    kept_reads = 0

    with pysam.AlignmentFile(source_bam, "rb") as src:
        with pysam.AlignmentFile(out_bam, "wb", header=src.header, threads=threads) as out:
            for read in src.fetch(region.contig, region.start, region.end):
                if read.is_unmapped:
                    continue

                total_reads += 1

                hash_frac = _xxh32_fraction(read.query_name, seed)
                if uniform_fraction is not None:
                    read_ratio = uniform_fraction
                else:
                    read_ratio = get_ratio(read)

                if hash_frac < read_ratio:
                    out.write(read)
                    kept_reads += 1

                if total_reads % 1_000_000 == 0:
                    pct = 100.0 * kept_reads / total_reads
                    log(f"[sample]   Processed {total_reads} reads, kept {kept_reads} ({pct:.1f}%)")

    pct = 100.0 * kept_reads / total_reads if total_reads else 0.0
    log(f"[sample] Processed {total_reads} reads, kept {kept_reads} ({pct:.1f}%)")

    # Sort and index
    # if not no_sort:
    #     log("[sample] Sorting output BAM...")
    #     tmp_sorted = out_bam + ".tmp.sorted.bam"
    #     try:
    #         pysam.sort("-o", tmp_sorted, out_bam)
    #         os.replace(tmp_sorted, out_bam)
    #         pysam.index(out_bam)
    #         log("[sample] Sorting and indexing complete.")
    #     except Exception as exc:
    #         log(f"Warning: Failed to sort/index output BAM: {exc}")

    log(f"[sample] Done. Output written to: {out_bam}")
    return 0
